src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1124 @@
+        const uint64_t hi_lo = __emulu(lhs >> 32, rhs);
+        const uint64_t lo_hi = __emulu(lhs, rhs >> 32);
+        const uint64_t hi_hi = __emulu(lhs >> 32, rhs >> 32);
+        const uint64_t cross = (lo_lo >> 32) + (hi_lo & 0xFFFFFFFF) + lo_hi;
+        hi = (hi_lo >> 32) + (cross >> 32) + hi_hi;
+        lo = (cross << 32) | (lo_lo & 0xFFFFFFFF);
+#endif
+
+        return lo ^ hi;
+    };
+
+    // If executed on little endian CPUs these 4 numbers will
+    // equal the first 32 byte of the original XXH3_kSecret.
+    static constexpr uint64_t XXH3_kSecret[4] = {
+        UINT64_C(0xbe4ba423396cfeb8),
+        UINT64_C(0x1cad21f72c81017c),
+        UINT64_C(0xdb979083e96dd4de),
+        UINT64_C(0x1f67b3b7a4a44072),
+    };
+
+    uint64_t inputs[4];
+    memcpy(&inputs[0], data, 32);
+
+#pragma warning(suppress : 26450) // Arithmetic overflow: '*' operation causes overflow at compile time. Use a wider type to store the operands (io.1).
+    uint64_t acc = dataSize * UINT64_C(0x9E3779B185EBCA87);
+    acc += XXH3_mul128_fold64(inputs[0] ^ XXH3_kSecret[0], inputs[1] ^ XXH3_kSecret[1]);
+    acc += XXH3_mul128_fold64(inputs[2] ^ XXH3_kSecret[2], inputs[3] ^ XXH3_kSecret[3]);
+    acc = acc ^ (acc >> 37);
+    acc *= UINT64_C(0x165667919E3779F9);
+    acc = acc ^ (acc >> 32);
+    return acc;
+}
+
+#pragma endregion
+
+[[nodiscard]] HRESULT AtlasEngine::_handleException(const wil::ResultException& exception) noexcept
+{
+    const auto hr = exception.GetErrorCode();
+    if (hr == DXGI_ERROR_DEVICE_REMOVED || hr == DXGI_ERROR_DEVICE_RESET || hr == D2DERR_RECREATE_TARGET)
+    {
+        _r = {};
+        WI_SetFlag(_invalidations, InvalidationFlags::device);
+        return E_PENDING; // Indicate a retry to the renderer
+    }
+    return hr;
+}
+
+void AtlasEngine::_createResources()
+{
+#ifdef NDEBUG
+    static constexpr
+#endif
+        // Why D3D11_CREATE_DEVICE_PREVENT_INTERNAL_THREADING_OPTIMIZATIONS:
+        // This flag prevents the driver from creating a large thread pool for things like shader computations
+        // that would be advantageous for games. For us this has only a minimal performance benefit,
+        // but comes with a large memory usage overhead. At the time of writing the Nvidia
+        // driver launches "+$cpu_thread_count more worker threads without this flag.
+        auto deviceFlags = D3D11_CREATE_DEVICE_SINGLETHREADED | D3D11_CREATE_DEVICE_PREVENT_INTERNAL_THREADING_OPTIMIZATIONS | D3D11_CREATE_DEVICE_BGRA_SUPPORT;
+
+#ifndef NDEBUG
+    // DXGI debug messages + enabling D3D11_CREATE_DEVICE_DEBUG if the Windows SDK was installed.
+    if (const wil::unique_hmodule module{ LoadLibraryExW(L"dxgidebug.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32) })
+    {
+        deviceFlags |= D3D11_CREATE_DEVICE_DEBUG;
+
+        const auto DXGIGetDebugInterface = reinterpret_cast<HRESULT(WINAPI*)(REFIID, void**)>(GetProcAddress(module.get(), "DXGIGetDebugInterface"));

```

Comment:
Why use D3D11 and not D3D12?

Reply:
Let me turn this question around: Why use D3D12 and not D3D11? Newer software isn’t faster than old software.
But hardware is and D3D12 allows you to better access these newer capabilities.
However this engine is so simple that the core rendering loop only needs 3 graphics commands per frame. There’s not much that the newer API could help with performance here if at all.
Last but not least, this engine is supposed to work on Windows 7 which only supports D3D10 (accessible through D3D11), because it is used in Visual Studio 2019 (and that one supports Windows 7 until 2029 - not kidding).

Reply:
Oh I just found out that there's a "D3D12onWin7" document!
But it's only for x64 and doesn't support WARP, which we need because we have to support DirectX 9 hardware. 😔

One day I'll probably check out D3D12 anyways, because it's interesting. 😄
I suspect that it won't improve performance, because - as far as I understand it - D3D12 helps parallelize things, whereas this engine doesn't have enough complexity to benefit from this. But maybe I'm wrong and in that case it'd be fun having a cool D3D12 engine for modern hardware.